File: C++_version/largest_number.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

bool compare(string a, string b) {
	if (a + b > b + a) {
		return true;
	} else {
		return false;
	}
}

class Solution {
public:
    string largestNumber(vector<int> &num) {
    	vector<string> num_str;
    	for (auto n : num) {
    		num_str.push_back(to_string(n));
    	}

        sort(begin(num_str), end(num_str), compare);

        string res;

        bool leading_zero = true;
        for(auto n : num_str) {
        	if (leading_zero) {
        		if (n != "0") {
        			leading_zero = false;
        			res += n;
        		} else {
        			continue;
        		}
        	} else {
        		res += n;
        	}
        }

        if (res == "") {
        	res = "0";
        }

        return res;
    }
};



int main()
{
	Solution sol;
	vector<int> num {3, 30, 34, 5, 9};
	string res = sol.largestNumber(num);
	cout << res << endl;

	return 0;
}

File: python_version/largest_number.py
"""
Given a list of non negative integers, arrange them such that they form the largest number.

For example, given [3, 30, 34, 5, 9], the largest formed number is 9534330.

Note: The result may be very large, so you need to return a string instead of an integer.
"""

class Solution:
    # @param num, a list of integers
    # @return a string
    def largestNumber(self, num):
        num = sorted([str(x) for x in num], cmp = self.compare)
        num.reverse()
        res = ''.join(num).lstrip('0')
        if res == '':
            res = '0'
        return res

    def compare(self, a, b):
        if a+b > b + a:
            return 1
        else:
            return -1






if __name__ == '__main__':
    num = [20, 1]
    num2 = [3, 30, 34, 5, 9]
    num3 = [121, 12]
    sol = Solution()
    print(sol.largestNumber(num2))
